pkg/policy/rule.go

```diff
@@ -93,6 +93,19 @@ func adjustL4PolicyIfNeeded(fromEndpoints []api.EndpointSelector, policy *L4Filt
 		// use a more permissive one
 		policy.FromEndpoints = nil
 	}
+
+	for _, newEp := range fromEndpoints {
+		found := false
+		for _, existing := range policy.FromEndpoints {
+			if newEp == existing {
```

Comment:
I doubt this does what you think. `api.FromEndpoints` is indeed comparable, but since its only field is a pointer, this will just check whether the two `api.FromEndpoints` contain the exact same pointer. This doesn't do a deep comparison, cf. the Golang spec: https://golang.org/ref/spec#Comparison_operators
* Struct values are comparable if all their fields are comparable. Two struct values are equal if their corresponding non-blank fields are equal.
* Pointer values are comparable. Two pointer values are equal if they point to the same variable or if both have value nil. Pointers to distinct zero-size variables may or may not be equal.

I would personally not worry about duplicates in `FromEndpoints`.

Reply:
Eek, that's a bit of a newbie gotcha. Thanks for the review.

I looked into how many of the tests would fail if we just appended here, and it looks like we just need this incremental to make everything pass, and then we could drop this deduplication:

```
diff --git a/pkg/policy/repository_test.go b/pkg/policy/repository_test.go
index c44f5b8c13ba..3244eb75c219 100644
--- a/pkg/policy/repository_test.go
+++ b/pkg/policy/repository_test.go
@@ -301,10 +301,26 @@ func (ds *PolicyTestSuite) TestMinikubeGettingStarted(c *C) {
        hash, err := selectorFromApp2[0].Hash()
        c.Assert(err, IsNil)

+       // Due to the lack of a set structure for L4Filter.FromEndpoints,
+       // merging multiple L3-dependent rules together will result in
+       // multiple instances of the EndpointSelector. This is just to get
+       // the comparison tests passing.
+       selectorFromApp2DupList := []api.EndpointSelector{
+               api.NewESFromLabels(
+                       labels.ParseSelectLabel("id=app2"),
+               ),
+               api.NewESFromLabels(
+                       labels.ParseSelectLabel("id=app2"),
+               ),
+               api.NewESFromLabels(
+                       labels.ParseSelectLabel("id=app2"),
+               ),
+       }
+
        expected := NewL4Policy()
        expected.Ingress["80/TCP"] = L4Filter{
                Port: 80, Protocol: api.ProtoTCP,
-               FromEndpoints: selectorFromApp2,
+               FromEndpoints: selectorFromApp2DupList,
                L7Parser:      "http",
                L7RulesPerEp: L7DataMap{
                        hash: api.L7Rules{
diff --git a/pkg/policy/rule.go b/pkg/policy/rule.go
index fb53bc2a0cca..6543ff8e41a2 100644
--- a/pkg/policy/rule.go
+++ b/pkg/policy/rule.go
@@ -94,18 +94,7 @@ func (policy *L4Filter) addFromEndpoints(fromEndpoints []api.EndpointSelector) b
                policy.FromEndpoints = nil
        }

-       for _, newEp := range fromEndpoints {
-               found := false
-               for _, existing := range policy.FromEndpoints {
-                       if newEp == existing {
-                               found = true
-                               break
-                       }
-               }
-               if !found {
-                       policy.FromEndpoints = append(policy.FromEndpoints, newEp)
-               }
-       }
+       policy.FromEndpoints = append(policy.FromEndpoints, fromEndpoints...)
        return false
 }
```